components/raftstore-v2/src/worker/tablet_gc.rs

```diff
@@ -107,27 +116,74 @@ impl<EK: KvEngine> Runner<EK> {
             logger,
             waiting_destroy_tasks: HashMap::default(),
             pending_destroy_tasks: Vec::new(),
+            background_pool: YatpPoolBuilder::new(DefaultTicker::default())
+                .name_prefix("tablet-gc-bg")
+                .thread_count(
+                    0,
+                    DEFAULT_BACKGROUND_POOL_SIZE,
+                    DEFAULT_BACKGROUND_POOL_SIZE,
+                )
+                .build_future_pool(),
         }
     }
 
-    fn trim(
-        tablet: &EK,
-        start_key: &[u8],
-        end_key: &[u8],
-        cb: Box<dyn FnOnce() + Send>,
-    ) -> engine_traits::Result<()> {
-        let start_key = keys::data_key(start_key);
-        let end_key = keys::data_end_key(end_key);
+    fn trim(&self, tablet: EK, start: Box<[u8]>, end: Box<[u8]>, cb: Box<dyn FnOnce() + Send>) {
+        let start_key = keys::data_key(&start);
+        let end_key = keys::data_end_key(&end);
         let range1 = Range::new(&[], &start_key);
         let range2 = Range::new(&end_key, keys::DATA_MAX_KEY);
-        tablet.delete_ranges_cfs(DeleteStrategy::DeleteFiles, &[range1, range2])?;
-        // TODO: Avoid this after compaction filter is ready.
-        tablet.delete_ranges_cfs(DeleteStrategy::DeleteByRange, &[range1, range2])?;
-        for r in [range1, range2] {
-            tablet.compact_range(Some(r.start_key), Some(r.end_key), false, 1)?;
+        // TODO: Avoid `DeleteByRange` after compaction filter is ready.
+        if let Err(e) = tablet
+            .delete_ranges_cfs(DeleteStrategy::DeleteFiles, &[range1, range2])
+            .and_then(|_| {
+                tablet.delete_ranges_cfs(DeleteStrategy::DeleteByRange, &[range1, range2])
+            })
+        {
+            error!(
+                self.logger,
+                "failed to trim tablet";
+                "start_key" => log_wrappers::Value::key(&start_key),
+                "end_key" => log_wrappers::Value::key(&end_key),
+                "err" => %e,
+            );
+            return;
         }
-        cb();
-        Ok(())
+        let logger = self.logger.clone();
+        self.background_pool
+            .spawn(async move {
+                let start_key = keys::data_key(&start);
```

Comment:
Why not just move `range1` and `range2` from outside?

Reply:
I don't get it. The way you write also creates a self-referential anonymous struct. Maybe you should also move `start_key` and `end_key` by calling drop explicitly in the closure.

Reply:
Or perhaps just delete L154 and L155 as `Range` is not 'static.

Reply:
I already explicitly move the variable when trying to fix this issue earlier.

> The way you write also creates a self-referential anonymous struct.

How? Only `start` and `end` is moved inside, no reference of them is moved inside.

Reply:
Since you are not using await, so there is no self-referential. There has nothing to do with self-referential problem, it's just the borrowed data can't be moved. That's why I said delete L154 and L155, so range is constructed again and start_key/end_key can be moved now as it's not borrowed anymore.

Reply:
What closure are you talking about? If you are referring to async move {}, then the generated struct is *not* a self re-reference struct. It reference `start_key` that is in function `trim`'s stack instead of closure fields. To generate a self reference struct, you need to use await, which will force store both the temporary variables `range1` and moved variables `start_key` in the same struct, hence it's self-referenced.

Reply:
You may assume the compiler is smart enough to generate something like:
```
let start_key = vec![];
let range1 = &start_key;

struct Closure<'a> {
    _start_key: Vec<u8>,
    _range1: &[u8],
}

let mut c = Closure {
    _start_key: start_key,
   _range1: &[],
};
c._range1 = &c._start_key;
```

But it's not. It only generates:
```
let start_key = vec![];
let range1 = &start_key;

struct Closure<'a> {
    _start_key: Vec<u8>,
    _range1: &[u8],
}

let mut c = Closure {
    _start_key: start_key,
   _range1: range1,
};
```

Reply:
I don't understand your "await" statement. In fact, if you consume the closure inside this function (which means having access to function's local context during the closure's lifetime, the closure can be either async or sync), there wouldn't be any self-reference at all, because the only in this case "It reference start_key that is in function trim's stack instead of closure fields".

And what I'm saying is, the rule of "you can't move a borrowed value" is simply the result of self-reference. The logic chain is (1) we want to move in temporary reference `range1`, (2) because it references another temporary value `start_key`, we want to move it that as well (3) voila, we got ourselves a self-referencing struct (4) so we must have the rule "you can't move a borrowed value".

I don't think your example below will work either. It can compile. But as soon as you try to move the `Closure` outside of function scope. It will err. Because the temporary address of `c` is referenced by `c._range1`, making `c` impossible to be moved out.

Reply:
> the rule of "you can't move a borrowed value" is simply the result of self-reference.

What about:
```
fn foo() {
    let v = vec![];
    let p = &v;
    bar(v);
}
```
?
You can't move a borrowed value because the value is borrowed, moved will change its ownership and create dangling reference.

The example I give is to explain how Rust desugar the closure, of course it won't compile.

For the self reference part, I suggest you to read docs about pin. Why it's created and how it works should answer your question.

Reply:
> You can't move a borrowed value because the value is borrowed, moved will change its ownership and create dangling reference.

This is a different context. It doesn't change the reasoning for this context. Rules may apply to different contexts. If it isn't necessary in one context, I'm sure rust will not impose it single-mindedly.

> The example I give is to explain how Rust desugar the closure, of course it won't compile.

Did you read my comment? I said "it can compile". The first code snippet can compile. But it wouldn't help this context because the `Closure` cannot be moved outside the function.

> For the self reference part, I suggest you to read docs about pin. Why it's created and how it works should answer your question.

I fail to comprehend the arrogance here. When I have already refuted your whole argument, and the only response you have is "go learn stuff"?

Reply:
The rule is part of ownership concept not specific to any particular context. I understand your comments completely, that's why I give an example in the first place to show you that the compiler just doesn't work the way you assume, the "logic chain" (the first example) is not how it work in reality (the second example is). The problem here is you don't understand my argument completely (and refuse to learn). Maybe I'm not the one who is arrogant here. :)

Reply:
Okay, let's list your arguments along the way:

(1) The way you write also creates a self-referential anonymous struct.

This is wrong. And I asked why and gave my reason.

(2) It reference start_key that is in function trim's stack instead of closure fields.

It certainly is not. The closure is moved into a thread pool. How can it still reference a temporal? My reasoning for this is done in the previous comment as well.

(3) To generate a self reference struct, you need to use await, which will force store both the temporary variables range1 and moved variables start_key in the same struct, hence it's self-referenced.

This is wrong. I tested personally and gave my logical reasoning as well.

Just to be more considerate here, if you are referring to making this `trim` function an async function returning a future of the closure, that would be a completely different situation and shouldn't be stated as "use await".

(4) You may assume the compiler is smart enough to generate something like: [...] I understand your comments completely, that's why I give an example in the first place to show you that the compiler just doesn't work the way you assume, the "logic chain" (the first example) is not how it work in reality (the second example is).

Well, first of all, the expression "You may assume the compiler is smart enough to [..]" can be understood as a snarky comment on a suboptimal compiler behavior. And this expression is common enough to let me believe you are suggesting that if the compiler would use the optimal solution, problem would disappear. Which is a faulty statement and that's why I commented my confusion.

After that, your comment explicitly says, "The example I give is to explain how Rust desugar the closure, of course it won't compile.". I think any seasoned English speaker can agree, you are referencing a previous context here, which is me saying it can't compile. But there's no such context. That's why I commented, are you reading my comment carefully? Are you misreading my "can compile" to "can't compile". Because otherwise this sentence is just too confusing to understand.

None of these misunderstandings is tech related. I believe moving forward, it is important to use English correctly and without ambiguity to avoid confusion.

(5) There has nothing to do with self-referential problem, it's just the borrowed data can't be moved.

Apparently, this is the core of our discussion here. Now I finally get to understand your argument, after the confusion in (4) about the code snippet is cleared. But I am still not convinced. Because the error message is not the same as if the compiler generated the code you gave, and it's not the same as you `fn foo()` example too. The error message is not "cannot move a borrowed value" and such, it is "`start_key`doesn't live long enough." This means compiler didn't generate the code you intended, instead it only moves in the reference (which is `range`), but didn't move in the value (which is `start_key`). And I put in the efforts to try to explain why compiler didn't move in the value (which is because that will generate a self-referencing struct).


> The problem here is you don't understand my argument completely (and refuse to learn). Maybe I'm not the one who is arrogant here. :)

Man, can I say the same thing. I believe I have done my due diligence to convey my meaning in my best efforts. And I came here in my best intention too. After all, I did say that I was only "learning about this today". I'm happy to change opinion If you give me anything concrete and correct to learn (and I did noticed my mistake about the redundant lines in (https://github.com/tikv/tikv/pull/14125#discussion_r1095482704). But no, you didn't gave any. The `Pin` docs have nothing to do here. It is a not-unsafe solution to get around the self-reference problem, but it does not explain if self-reference is the fundamental obstacle of writing the code here without redundant initialization.

Reply:
discussed offline. Some consensus here: (2)(4)(5) -> the chain actually has a third step to function: "move reference `range`" -> "move value `start_key` -> "alter reference `range` to point to `start_key` at a new location". The 3rd step violates the "move cannot change value" principle. So it shouldn't be generated by the compiler. If we don't allow the 3rd step, then the 2nd step won't work because its location is still being reference. (1)(3) -> "use await" refers to using await inside the closure, which forces saving both `range` and `start_key`. For normal cases (without await), even though both value and reference are accessible in the closure, a self-referencing struct can be optimized away.